core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactory.js

```diff
@@ -0,0 +1,210 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Factory for creating instances of NumberWithUnits and Units
+ * domain objects.
+ */
+
+oppia.constant('NUMBER_WITH_UNITS_PARSING_ERRORS', {
+  INVALID_VALUE:
+    'Please ensure that value is either a fraction or a number',
+  INVALID_CURRENCY:
+    'Please enter a valid currency (e.g., $ 5)',
+  INVALID_UNIT_CHARS:
+    'Please ensure that unit only contains numbers, alphabets, (, ), *, ^, /, -'
+});
+
+oppia.factory('NumberWithUnitsObjectFactory', [
+  'UnitsObjectFactory', 'FractionObjectFactory',
+  'NUMBER_WITH_UNITS_PARSING_ERRORS', function(
+      UnitsObjectFactory, FractionObjectFactory,
+      NUMBER_WITH_UNITS_PARSING_ERRORS) {
+    var NumberWithUnits = function(type, real, Fraction, Units) {
+      this.type = type;
+      this.real = real;
+      this.Fraction = Fraction;
+      this.Units = Units;
+    };
+
+    NumberWithUnits.fromRawInputString = function(rawInput) {
+      rawInput = rawInput.trim();
+      var type = '';
+      var real = '';
+      var Fraction = '';
+      var units = '';
+      var value = 0;
+      var Units = '';
+
+      // Start with digit when there is no currency unit.
+      if (rawInput.match(/^\d/)) {
+        var ind = rawInput.indexOf(' ');
+        if (ind === -1) {
+          // There is value with no units.
+          value = rawInput;
+        } else {
+          value = rawInput.substr(0, ind);
+          units = rawInput.substr(ind + 1).trim();
+        }
+      } else {
+        var ind = rawInput.indexOf(' ');
+        if (ind === -1) {
+          throw new Error(NUMBER_WITH_UNITS_PARSING_ERRORS.INVALID_CURRENCY);
+        }
+        units = rawInput.substr(0, ind) + ' ';
+        var ind2 = rawInput.indexOf(' ', ind + 1);
+        if (ind2 !== -1) {
+          value = rawInput.substr(ind + 1, ind2 - ind - 1).trim();
+          units += rawInput.substr(ind2 + 1).trim();
+        } else {
+          value = rawInput.substr(ind + 1).trim();
+          units = units.trim();
+        }
+      }
+      if (value.match(/[a-z]/i) || value.match(/[*^$()]/)) {
+        throw new Error(NUMBER_WITH_UNITS_PARSING_ERRORS.INVALID_VALUE);
+      }
+
+      if (value.includes('/')) {
+        type = 'fraction';
+        Fraction = FractionObjectFactory.fromRawInputString(value);
+      } else {
+        type = 'real';
+        real = parseFloat(value);
+      }
+      if (units !== '') {
+        if (units.match(/[^0-9a-z/* ^$-]/i)) {
```

Comment:
Add comments whenever you use regex. It will be easier to understand for other programmers that way.

Reply:
Done.